### PR TITLE
feat(pubsub): subscription callback concurrency control

### LINKED ISSUE #4652: Optionally run subscription callbacks out of order
Once we have implemented flow control for subscriptions (see #4645) we should consider running subscription callbacks for a large `Pull` response in parallel.

### PR DESCRIPTION
**Pull request by coryan**
Add support for subscriptions to *schedule* multiple callbacks in
parallel.  Note that the number of callbacks that *run* in parallel
depends on the number of threads servicing the completion queue
associated with the subscription. The previous behavior (only one
callback at a time) remains possible via configuration in the
`pubsub::SubscriptionOptions`. The doxygen documentation for
`pubsub::SubscriptionOptions` has more details.

Fixes #4652 

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4938)

### CODE REVIEW

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+namespace {
+
+using ::testing::AtLeast;
+
+/// @test Verify session can schedule parallel callbacks.
+TEST(SubscriptionConcurrentControlTest, ScheduleParallelCallbacks) {
+  auto mock = std::make_shared<pubsub_testing::MockSubscriberStub>();
+  pubsub::Subscription const subscription("test-project", "test-subscription");
+
+  std::mutex mu;
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:36]: `#include <mutex>`

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      });
+  EXPECT_CALL(*mock, AsyncModifyAckDeadline)
+      .WillRepeatedly([](google::cloud::CompletionQueue&,
+                         std::unique_ptr<grpc::ClientContext>,
+                         google::pubsub::v1::ModifyAckDeadlineRequest const&) {
+        return make_ready_future(Status{});
+      });
+
+  google::cloud::CompletionQueue cq;
+  std::vector<std::thread> tasks;
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:75]:
```
#include <thread>
#include <vector>
```

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  std::generate_n(std::back_inserter(tasks), 4,
+                  [&] { return std::thread([&cq] { cq.Run(); }); });
+
+  std::mutex handlers_mu;
+  std::condition_variable handlers_cv;
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:80]: `#include <condition_variable>`

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  std::deque<pubsub::AckHandler> handlers;
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:81]: `#include <deque>`

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+TEST(SubscriptionConcurrentControlTest, ScheduleParallelCallbacks) {
+  auto mock = std::make_shared<pubsub_testing::MockSubscriberStub>();
+  pubsub::Subscription const subscription("test-project", "test-subscription");
+
+  std::mutex mu;
+  int count = 0;
+  auto constexpr kBatchSize = 8;
+  auto constexpr kTestHwm = 4;
+  static_assert(kBatchSize % kTestHwm == 0,
+                "batch size must be a multiple of hwm for this test");
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:41]: So, set `kBatchSize = N * kTestHwm`?

File: google/cloud/pubsub/internal/watermark_flow_control.cc [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  std::lock_guard<std::mutex> lk(mu_);
+  if (overflow_ || current_ >= hwm_) return false;
+  current_ += size;
+  if (current_ >= hwm_) overflow_ = true;
+  return true;
+}
+
+bool WatermarkFlowControl::Release(std::size_t size) {
+  std::lock_guard<std::mutex> lk(mu_);
+  current_ -= size;
```
Comment (devbww) [inline google/cloud/pubsub/internal/watermark_flow_control.cc:32]: Limit `size` to `current_` somehow?

File: google/cloud/pubsub/subscription_options.h [hunk 7]
```diff
@@ -55,8 +55,42 @@ class SubscriptionOptions {
...
   // TODO(#4645) - add options for flow control
 
+  /**
+   * Set the high watermark and low watermarks for callback concurrency.
+   *
+   * The Cloud Pub/Sub C++ client library will schedule parallel callbacks as
+   * long as the number of outstanding callbacks is less than the high
+   * watermark. Once the watermark is reached the client will not resume
+   * scheduling callbacks until the number of outstanding callbacks is at or
+   * below the low watermark. Using hysteresis prevents and instability.
```
Comment (devbww) [inline google/cloud/pubsub/subscription_options.h:65]: `s/and/any/` or `s/ and//`.

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_concurrency_control_test.cc]: Doh, done.

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:41]: Done.

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:78]: Done

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:83]: Done

File: google/cloud/pubsub/internal/subscription_concurrency_control_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_concurrency_control_test.cc:84]: Done.

File: google/cloud/pubsub/internal/watermark_flow_control.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/pubsub/internal/watermark_flow_control.cc]: Done.

File: google/cloud/pubsub/subscription_options.h (hunk 7 above)
Comment (coryan) [reply google/cloud/pubsub/subscription_options.h]: Done.

**Pull request merged**